src/ma/upsampler/upsample_supply_monthly_gen_to_hh.py

```
@@ -0,0 +1,309 @@
+
+    # Group by year and month, sum only numeric columns
+    numeric_cols = grid_mix_monthly.select_dtypes(include="number").columns.tolist()
+    numeric_cols.remove("year")
+    numeric_cols.remove("month")
+
+    # Group by year and month, sum only numeric columns
+    grid_mix_monthly = grid_mix_monthly.groupby(["year", "month"])[numeric_cols].sum()
+    # Reset index to have year and month as columns
+    grid_mix_monthly = grid_mix_monthly.reset_index()
+
+    return grid_mix_monthly
+
+
+def _calculate_scaling_factors(
+    grid_mix_by_tech_by_month: pd.DataFrame, supply_gen_by_month: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Calculate scaling factors for each technology, supplier and month using vectorized operations.
+
+    Takes prepared grid mix and supplier data and calculates the proportion
+    of grid generation that should be allocated to each supplier.
+    """
+    grid_mix = grid_mix_by_tech_by_month.copy()
+    tech_columns = [col for col in grid_mix.columns if col.endswith("_mwh")]
+
+    # Create a copy of the supplier data to avoid modifying the original
+    supply_gen = supply_gen_by_month.copy()
+    supply_gen = supply_gen.rename(columns={"month_num": "month"})  # required for joining
+
+    # Select only the technology columns and year/month
+    tech_data = grid_mix[["year", "month"] + tech_columns].copy()
+
+    # Convert from wide to long format
+    grid_mix_long = pd.melt(
+        tech_data,
+        id_vars=["year", "month"],
+        value_vars=tech_columns,
+        var_name="tech_column",
+        value_name="grid_total_mwh",
+    )
+
+    # Standardise tech names
+    grid_mix_long["tech"] = grid_mix_long["tech_column"].str.replace("_mwh", "")
+
+    # Merge with supplier data
+    merged_data = pd.merge(
+        supply_gen[["year", "month", "tech", "current_holder", "rego_mwh"]],
+        grid_mix_long[["year", "month", "tech", "grid_total_mwh"]],
+        on=["year", "month", "tech"],
+        how="inner",
```

_left_ would be more robust because, where we're missing `grid_mix` entries, it doesn't silently drop `supply`.

---

(This would be a good thing to test)

---

I've added a test for the new `expected_periods` check discussed in the comment below, which confirms we'll raise a value err if there are gaps in the HH data. 

Is that what you intended? Or do you mean it would be good to test the merge somehow?

---

Perhaps redundant now we have the check on `expected_periods` but setting `how="left"` would give a row for each `(year, month, tech)` there was _supply_, even if there was no associated entry in _grid_mix_.

---

P.s. looks like this file got dropped in the latest commit

---

Re-added (post-rename)!